### ISSUE TITLE
Overriding Signature at call site of variadic function programmatically from script 

### ISSUE THREAD

**Issue by pabx06** [NationalSecurityAgency/ghidra#2411]
hello.
i have over 10k callSite  of a printf like function.
I would like to 'Override Signature' programmatically and systematically from a GhidraScript. So i can have proper types inferred.

i could not figure out witch api to invoke from:  https://github.com/NationalSecurityAgency/ghidra/blob/aa299897c6b84e16ecf228d82cf8957a9529b819/Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/decompile/actions/OverridePrototypeAction.java#L35

**Comment by astrelsky** [comment]
If the architecture isn't stack based you can give this a shot. 

https://www.github.com/astrelsky/ghidra_scripts/tree/master/PrintfSigOverrider.java

**Comment by astrelsky** [comment]
I don't think you need to commit the local parameters. The override should be enough to influence the type propagation algorithm.

**Comment by pabx06** [comment]
About the **SymbolicPropogator**.
is there an API for getting a list of all possible getRegisterValue of all Possible Control flows ?

```
	/**
	 * Process a subroutine using the processor function.
	 * The process function can control what flows are followed and when to stop.
	 * 
	 * @param startAddr start address
	 * @param restrictSet the address set to restrict the constant flow to
	 * @param eval the context evaluator to use
	 * @param saveContext true if the context should be saved
	 * @param monitor the task monitor
	 * @return the address set of instructions that were followed
	 * @throws CancelledException if the task is cancelled
	 */
	public AddressSet flowConstants(Address startAddr, AddressSetView restrictSet,
			ContextEvaluator eval, boolean saveContext, TaskMonitor monitor)
```

```
	/**
	 * Get constant or register relative value assigned to the 
	 * specified register at the specified address
	 * @param toAddr address
	 * @param reg register
	 * @return register value
	 */
	public Value getRegisterValue(Address toAddr, Register reg) 
```

ie:
```
funct(){
 //(stack var)
 char * format ; 
 if(global_RW_var == 0){
   format = "%s %s"
 }
 else{
   format = "%d"
 }
 //r0 could have two different values.
 printf(format,...) 
}
```

as for the CFG found: #444

**Comment by pabx06** [comment]
override few calls within the script : a couple of thing didn't make much sense :
## context action missing
- 'Remove Signature Overide' context action from the decomp not present
![image](https://user-images.githubusercontent.com/4013370/97781434-a22a0900-1b8b-11eb-85b8-74e49251754c.png)

## Signedness
The decomp discard randomly Signedness some time yes some time no. or does the ovveride get realy applied ?

ie:
- String is %4d,%02d,%02d,%2d,%02d 
- Parsed to  int,int,int,int,int 
- FunctionDefinition applied is void tmpname(int LEVEL, char * format, **int , int , int , int , int** )

but when i examine with right click `Override Signature` decompiler context menu  in the decomp show me that the signature is : void log_status(int,char *,**uint,uint,uint,uint,uint**) . as if it completely discarding Signedness or not applying . So in the pseudo C code i had few params with 0xFFFFFFFF instead of -1

Another example where the FunctionDefinition applied is OK as shown in the bookmarks.
but decompiler display something else : 
`%04x => uint => void tmpname(int LEVEL, char * format, uint )`
![image](https://user-images.githubusercontent.com/4013370/97781823-2d0c0300-1b8e-11eb-87f1-5ab027c12c22.png)
despite that a label `log_status::override::prt_4019bce0_3f00635e`

**Comment by astrelsky** [comment]
The script I had written was specifically for printf and expects the format string to be the first argument. There is no override present in the first screenshot. I think the action for overriding a function signature attempts to use the currently assumed types by the decompiler as a convenience. The best way to find the unit tests would be to look for the tests for the decompiler plugin and the override action.

**Comment by pabx06** [comment]
> The script I had written was specifically for printf and expects the format string to be the first argument. There is no override present in the first screenshot. I think the action for overriding a function signature attempts to use the currently assumed types by the decompiler as a convenience. The best way to find the unit tests would be to look for the tests for the decompiler plugin and the override action. The action class is `OverridePrototypeAction`, it doesn't appear to be covered by any unit tests and I don't have the ability to set up the correct Eclipse dev environment with the tools to show otherwise.
> 
> This is the epitome of something that was never intended to be done in a ghidra script.

I think i know now. I spoted the magic api call

https://github.com/NationalSecurityAgency/ghidra/blob/aa299897c6b84e16ecf228d82cf8957a9529b819/Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/decompile/actions/OverridePrototypeAction.java#L271
